File: python/src/algorithm/geometry/planes/quadrilaterals/square.py
from typing import final

from algorithm.geometry import PlaneFigure
from algorithm.utils import validation


@final
class Square(PlaneFigure):
    """
    Square
    ======

    A regular quadrilateral with four equal sides and four right angles (90°).
    It has four lines of symmetry and equal diagonals that bisect each other.

    Example
    -------
    ```
    from algorithm.geometry.planes.quadrilaterals import Square

    square: Square = Square(5)

    print(f"Square Perimeter: {square.perimeter()}")
    print(f"Square Area: {square.area()}")
    ```
    """

    def __init__(self, side: float):
        validation.validate_positive(side, "side")
        self._side = side

    @property
    def side(self) -> float:
        """
        Returns the side length of the square.
        """
        return self._side

    @property
    def diagonal(self) -> float:
        """
        Returns the length of the diagonal of the square.
        """
        return pow(2, 0.5) * self.side

    def perimeter(self) -> float:
        return 4 * self.side

    def area(self) -> float:
        return pow(self.side, 2)

File: java/src/main/java/algorithm/geometry/planes/Square.java
package algorithm.geometry.planes;

import algorithm.geometry.Plane;
import algorithm.utils.Validation;

public final class Square implements Plane {
    private final double side;

    public Square(double side) throws Exception {
        Validation.validatePositive(side, "side");
        this.side = side;
    }

    public double getSide() {
        return side;
    }

    @Override
    public double perimeter() {
        return 4 * side;
    }

    @Override
    public double area() {
        return Math.pow(side, 2);
    }

}
